### PR TITLE
Add badges to mercator

### PR DESCRIPTION
**Pull request by MagdaN**
To add badges run:
bin/import_resources ./etc/development.ini src/adhocracy_mercator/adhocracy_mercator/scripts/sample_badge.json 

To add badge to proposal run:
bin/add_badge_assignment ./etc/development.ini userpath badgepath proposalversionpath proposalitempath

or use a json (I created a sample in adhocracy_mercator/scripts/)

bin/add_badge_assignment_from_json  ./etc/development.ini jsonfile

### CODE REVIEW

**Comment by joka** [comment]
zwei python unit test schlagen fehl

**Comment by joka** [comment]
using optparse is deprecated, can you use argpars? god example for scripts is here:
      src/adhocracy_mercator/adhocracy_mercator/scripts/export_users.py

**Comment by xi** [comment]
> using optparse is deprecated, can you use argpars? god example for scripts is here:
> src/adhocracy_mercator/adhocracy_mercator/scripts/export_users.py

I created the related issue #1271.

**Comment by MagdaN** [comment]
@joka @xi changed it - can you check it. There code also contains frontend code.

**Comment by xi** [comment]
Frontend looks good to me. I did some gardening/refactoring in db5cc9e/cc3849d and fixed a conflict with master in 86f9c3a.

File: src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+from adhocracy_core.resources.badge import IBadgeAssignment as BadgeResource
+from adhocracy_core.sheets.badge import IBadgeAssignment as BadgeSheet
+
+
+def add_badge_assignment():
+    """add badgeassignment.
+
+    usage::
+      bin/create_badge <config> <user> <badge> <proposal> <parent>
+    """
```
Comment (joka) [inline src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py]: Do we need this script? for me add_badge_assignment_from_json is enough

File: src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    root = env['root']
+    registry = env['registry']
+
+    user = args.user
+    badge = args.badge
+    proposal_version = args.proposal
+    parent = args.parent
+
+    user = find_resource(root, user)
```
Comment (joka) [inline src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py]: you override variables with different values here. 
 I would rename (user_path and user) theme or better make it shorter (user = find_resource(root, args.user)

File: src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    badge = find_resource(root, badge)
+    proposal_version = find_resource(root, proposal_version)
+    parent = find_resource(root, parent)
+
```
Comment (joka) [inline src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py]: public functions should be short around 5 lines and have one abstraction level to improve readability.
I would move the code that is creating the assignment to another function. This way you we can also reuse it and be more DRY.

File: src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                  {'subject': user,
+                   'badge': badge,
+                   'object': proposal_version}}
+
+    registry.content.create(BadgeResource.__identifier__,
+                            parent=parent,
+                            appstructs=appstructs)
+
+    transaction.commit()
+
```
Comment (joka) [inline src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py]: `env['closer']()` is missing here

File: src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+import inspect
+import argparse
+import transaction
+import json
+
+from pyramid.paster import bootstrap
+from pyramid.traversal import find_resource
+from adhocracy_core.resources.badge import IBadgeAssignment as BadgeResource
+from adhocracy_core.sheets.badge import IBadgeAssignment as BadgeSheet
+
```
Comment (joka) [inline src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py]: I was confused what BadgeSheet is exactly.
To seperate resource.badge.IBadgeAssignment and sheets.badge.IBadgeAssignment we import the full namesspace for the sheet: import adhocracy_core.sheets.badge.IBadgeAssignement

File: src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                      {'subject': user,
+                       'badge': badge,
+                       'object': proposal_version,
+                       'description': description}}
+
+        registry.content.create(BadgeResource.__identifier__,
+                                parent=parent,
+                                appstructs=appstructs)
+
+    transaction.commit()
```
Comment (joka) [inline src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py]: env['closer']() is missing here

File: src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+
+def _load_json(filename: str) -> [dict]:
+    with open(filename, 'r') as f:
```
Comment (joka) [inline src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py]: this helper function can move to adhocracy_core.utils, we need this multiple times.

File: src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    proposal_version = find_resource(root, proposal_version)
+    parent = find_resource(root, parent)
+
+    appstructs = {BadgeSheet.__identifier__:
+                  {'subject': user,
+                   'badge': badge,
+                   'object': proposal_version}}
+
+    registry.content.create(BadgeResource.__identifier__,
+                            parent=parent,
```
Comment (joka) [inline src/adhocracy_mercator/adhocracy_mercator/scripts/manage_badges.py]: BadgeAssignments have the be created in the post_pool 
defined in the badgeable sheet.
If don't want to use the sheet you can use `find_service(proposal, 'badge_assignments')`

File: src/adhocracy_mercator/adhocracy_mercator/scripts/sample_badge.json [hunk 9]
```diff
@@ -0,0 +1,24 @@
+[
```
Comment (joka) [inline src/adhocracy_mercator/adhocracy_mercator/scripts/sample_badge.json]: normally we do test driven development and 100% unit/integration test coverage for all backend code.
I would be good if we also start doing this for mercator scripts instead manual testing with sample scripts. An example is here: 
https://github.com/liqd/adhocracy3.mercator/blob/master/src/adhocracy_core/adhocracy_core/scripts/test_import_users.py

If setting up an intergration test is to trouble some we could also make it a functional test

**Comment by joka** [comment]
I did some comment mostly regarding code style.
If we don't have time for this, we should a least use the right post_pool to create badge assignments
and add env['close'r]()

File: src/adhocracy_core/adhocracy_core/sheets/badge.py [hunk 10]
```diff
@@ -138,6 +138,7 @@ class BadgeAssignmentSchema(colander.MappingSchema):
     subject = Reference(reftype=BadgeSubjectReference)
     badge = Reference(reftype=BadgeReference)
     object = Reference(reftype=BadgeObjectReference)
+    description = Text()
```
Comment (nidico) [inline src/adhocracy_core/adhocracy_core/sheets/badge.py]: This isn't needed as it already comes with the `IDescription` sheet which you added to the resource above.

(Btw this was also wrong with the `IBadge` resource, but fixed with #1270).

File: src/adhocracy_core/adhocracy_core/sheets/test_badge.py [hunk 11]
```diff
@@ -89,7 +89,8 @@ def test_get_empty(self, meta, context):
         inst = meta.sheet_class(meta, context)
         assert inst.get() == {'subject': None,
                               'badge': None,
-                              'object': None}
+                              'object': None,
+                              'description' : ''}
```
Comment (nidico) [inline src/adhocracy_core/adhocracy_core/sheets/test_badge.py]: same

**Comment by MagdaN** [comment]
updated script @pallix @joka could you please look at it and merge soon.

**Comment by MagdaN** [comment]
@pallix: what do you mean by 'Also other scripts can be refactored to use it'? I really need this script to keep on working in the frontend.

**Comment by MagdaN** [comment]
@pallix @joka: I added the test. Please can that be merged now..

**Pull request merged**